### ISSUE TITLE
New end-to-end regression test module

### ISSUE THREAD

**Issue by lfcnassif** [sepinf-inc/IPED#1735]
This is a very old idea. We could have a new `iped-tests` module to run full processing on some small public images (downloaded on demand and cached, or maybe copied to this repo) and check some basic processing results, failing the build if not good:
- Number of allocated files, deleted (non-carved) files, file paths, MACB times, content hashes;
- Number of file signatures recognized per per mime-type and per category;
- Number of carved files per mime-type and their hashes;
- Extracted metadata (exif, office, emails headers, etc) keys and values for some (or all since this will be automated) files;
- Numer of thumbnails generated per format, parsing exceptions per format, and other default filters counts;
- Number of search results per format after searching for common words: prepositions, conjunctions, nouns. This evaluates content parsing and indexing at the same time;

I usually run most of above tests manually on some real world evidences before each major or feature release, and basic tests before fixing releases. That consumes a lot of time...

Running those tests on small public standard images automatically and failing the build if results are not OK should help a lot to detect unintentional regressions, including when we upgrade critical dependencies (sleuthkit, tika, lucene).

Additionally, we can produce diff reports of above stats between two different versions (a stable one and a release candidate) after running on real world regression datasets to be informed by the developer/tester. This usually shows divergences between versions that should be checked manually, because they can be bad or good: some files can move from one category to a better one, less carved files because more garbage is discarded or because of categorization differences, less search results because of less parsing exceptions (causing strings parser to run on less files)...

Maybe our new intern can help us to bring this old wish to life.

**Comment by lfcnassif** [comment]
The diff reports actually is an idea borrowed from Tika project and maybe we can use a similar approach:
https://cwiki.apache.org/confluence/display/tika/TikaEval

**Comment by lfcnassif** [comment]
- Number of subitems extracted per parent container, their hashes and metadata (from the container like MACB)

**Comment by lfcnassif** [comment]
@marcus6n, after some regressions in past releases and after the regression I introduced in 4.2.0 (#2484), I would like you to help with this. It can be implemented as a new IPED module, like iped-tests, or as a different project, in which case it could be implemented in java or python. What do you think?

**Comment by marcus6n** [comment]
@lfcnassif I'll try to implement this module! I think the proposal is great, and I believe it will add a lot to the version validation process. I'll start structuring something, and if any difficulties arise along the way, I’d appreciate your help in aligning the approaches.

**Comment by marcus6n** [comment]
@lfcnassif So far, I’ve completed the following steps for automating the test image processing:

- Implemented the download of a small public `.E01` test image (`ntfs1-gen1.E01`) if not already cached.

- Cached the downloaded image using GitHub Actions to avoid re-downloading on each build.

- Created a processing class that locates the IPED executable (`iped.exe`) in the release folder and runs it automatically on the test image using ProcessBuilder.

- Confirmed that the image is processed successfully and results are stored in a predefined output directory (`iped-tests/target/iped-result`).

The next step will be to implement automated validations on the output to detect regressions or processing failures and make the build fail accordingly.

**Comment by marcus6n** [comment]
The test failed on GitHub Actions because it tried to execute a Windows .exe on a Linux runner, which isn't supported. However, the test runs successfully on Windows environments. Is there a way to configure this to run properly in CI?

**Comment by lfcnassif** [comment]
> Is there a way to configure this to run properly in CI?

Yes, you need to change your code. Instead of calling iped.exe directly, you should check if OS is Windows or Linux, if the latter, call `java -jar iped.jar` instead.

**Comment by lfcnassif** [comment]
> [@lfcnassif](https://github.com/lfcnassif) So far, I’ve completed the following steps for automating the test image processing:
> 
> * Implemented the download of a small public `.E01` test image (`ntfs1-gen1.E01`) if not already cached.
> * Cached the downloaded image using GitHub Actions to avoid re-downloading on each build.
> * Created a processing class that locates the IPED executable (`iped.exe`) in the release folder and runs it automatically on the test image using ProcessBuilder.
> * Confirmed that the image is processed successfully and results are stored in a predefined output directory (`iped-tests/target/iped-result`).
> 
> The next step will be to implement automated validations on the output to detect regressions or processing failures and make the build fail accordingly.

Great, thanks @marcus6n!

**Comment by lfcnassif** [comment]
@marcus6n, I saw you are pushing commits to branch iped-tests. But please, open a pull request, in draft mode if it is not finished, so we can better follow your progress on this, thanks.

**Comment by marcus6n** [comment]
Sure, @lfcnassif! I’ll open a draft pull request so you can better follow the progress and provide feedback during the development. That should make it easier to review and align the next steps.
